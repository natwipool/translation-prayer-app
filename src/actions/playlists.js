import uuid from 'uuid';

// ADD_PLAYLIST
export const addPlaylist = ({ name = '', lists = [] } = {}) => ({
  type: 'ADD_PLAYLIST',
  playlist: {
    id: uuid(),
    name,
    lists
  }
});

// REMOVE_PLAYLIST
export const removePlaylist = ({ id } = {}) => ({
  type: 'REMOVE_PLAYLIST',
  id
});

// EDIT_PLAYLIST
export const editPlaylist = (id, updates) => ({
  type: 'EDIT_PLAYLIST',
  id,
  updates
});
