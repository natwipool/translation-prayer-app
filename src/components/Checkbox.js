import React from 'react';

// class Checkbox extends React.Component {
//   constructor(props) {
//     super(props)

//     this.state = {
//       checked: props.lists.find((list) => list === props.precept)
//     }
//   }

//   onAddLists = (list) => {
//     this.setState(() => ({
//       checked: !this.state.checked
//     }))
//     console.log(this.state.checked)
//     this.props.onAddLists(list)
//   }
  
//   onRemoveLists = (list) => {
//     this.setState(() => ({
//       checked: !this.state.checked
//     }))
//     this.props.onRemoveLists(list); 
//   }

//   render() {
//     return (
//       <div>
//         <input
//           type="checkbox"
//           name={this.props.precept}
//           value={this.props.precept}
//           checked={this.state.checked}
//           onChange={e => {
//             const list = e.target.value;
//             if (e.target.checked) {
//               this.onAddLists(list)  
//             } else if (!e.target.checked) {
//               this.onRemoveLists(list);  
//             }
//           }}
//         />
//         <label htmlFor={this.props.precept}>
//           {this.props.precept}
//           {
//             this.props.lists.indexOf(this.props.precept) > -1 && (
//               this.props.lists.indexOf(this.props.precept) + 1
//             )
//           }
//         </label>
//       </div>
//     )
//   }
// }

const Checkbox = (props) => (
  <div>
    <input
      type="checkbox"
      name={props.precept}
      value={props.precept}
      checked={props.lists.find((list) => list === props.precept)}
      onChange={e => {
        const list = e.target.value;
        if (e.target.checked) {
          props.onAddLists(list)  
        } else if (!e.target.checked) {
          props.onRemoveLists(list);  
        }
      }}
    />
    <label htmlFor={props.precept}>
      {props.precept}
      {
        props.lists.indexOf(props.precept) > -1 && (
          props.lists.indexOf(props.precept) + 1
        )
      }
    </label>
  </div>
);

export default Checkbox;
