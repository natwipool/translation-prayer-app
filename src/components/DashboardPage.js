import React from 'react';
import PrefacePage from './PrefacePage'

const DashboardPage = () => (
  <div>
    <PrefacePage />
  </div>
);

export default DashboardPage;