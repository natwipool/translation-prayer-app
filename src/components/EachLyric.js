import React from 'react';

const EachLyric = (props) => (
  <div>
    {props.lyric}
  </div>
);

export default EachLyric;