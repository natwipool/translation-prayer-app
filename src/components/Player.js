import React from 'react';
import ReactHowler from 'react-howler';
import raf from 'raf';
import secToMin from 'sec-to-min';

export class Player extends React.Component {
  constructor(props) {
    super(props);

    this.sources = this.props.playlists.map(
      playlist => playlist.filename.map((name) => (
        `/audio/${name}`
      ))
    );
    console.log(this.sources)
    this.state = {
      isPlaying: false,
      currentIndex: 0,
      loaded: false,
      seek: undefined,
      duration: undefined
    };
  }
  componentDidMount() {
    this.props.onRef(this);
  }
  componentWillUnmount() {
    this.clearRAF();
  }

  handleToggle = () => {
    this.setState(() => ({ isPlaying: !this.state.isPlaying }));
    this.props.handlePlayButton();
  };

  handleOnLoad = () => {
    this.setState(() => ({
      loaded: true,
      duration: this.player.duration()
    }));
  };

  handleOnPlay = () => {
    this.renderSeekPos();
  };

  handleStop = () => {
    this.player.stop();
    this.setState(() => ({
      isPlaying: false
    }));
    this.renderSeekPos();
    this.props.handleStopButton();
  };

  handleNext = () => {
    this.clearRAF();
    if (this.state.currentIndex < this.sources.length - 1) {
      this.setState(prevState => ({
        currentIndex: prevState.currentIndex + 1
      }));
      this.props.handlePlayNext();
    }
  };

  handlePrev = () => {
    this.clearRAF();
    if (this.state.currentIndex > 0) {
      this.setState(prevState => ({
        currentIndex: prevState.currentIndex - 1
      }));
      this.props.handlePlayPrev();
    }
  };

  handleAutoPlayNext = () => {
    this.clearRAF();
    
    if (this.state.currentIndex >= (this.sources.length - 1)) {
      this.handleStop();
    } else {
      this.setState(prevState => ({
        isPlaying: true,
        currentIndex: prevState.currentIndex + 1
      }));
      this.props.handlePlayNext();
    }  
  };

  playEachPrecept = playIndex => {
    this.clearRAF();   
    this.setState(() => ({
      isPlaying: true,
      currentIndex: playIndex
    }));
  };

  renderSeekPos = () => {
    this.setState(() => ({
      seek: this.player.seek()
    }));
    if (this.state.isPlaying) {
      this._raf = raf(this.renderSeekPos);
    }
  };

  clearRAF = () => {
    raf.cancel(this._raf);
  };

  render() {
    return (
      <div>
        <p>{this.state.loaded ? this.props.playlists[this.state.currentIndex].precept : 'Loading'}</p>
        <p>
          {'Status: '}
          {this.state.seek !== undefined ? secToMin(this.state.seek) : '0.00'}
          {' / '}
          {this.state.duration ? secToMin(this.state.duration) : '0.00'}
        </p>

        <ReactHowler
          src={this.sources[this.state.currentIndex]}
          playing={this.state.isPlaying}
          onLoad={this.handleOnLoad}
          onPlay={this.handleOnPlay}
          onEnd={() => {
            this.handleAutoPlayNext();
          }}
          ref={ref => (this.player = ref)}
        />
        <button
          onClick={this.handlePrev}
          disabled={this.state.currentIndex <= 0}
        >
          Prve
        </button>
        <button onClick={this.handleToggle}>
          {this.state.isPlaying ? 'Pause' : 'Play'}
        </button>
        <button onClick={this.handleStop}>Stop</button>
        <button
          onClick={this.handleNext}
          disabled={this.state.currentIndex >= this.sources.length - 1}
        >
          Next
        </button>
      </div>
    );
  }
}

export default Player;
