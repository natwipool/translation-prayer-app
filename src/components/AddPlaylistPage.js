import React from 'react';
import { connect } from 'react-redux';
import PlaylistForm from './PlaylistForm';
import { addPlaylist } from '../actions/playlists';

export class AddPlaylistPage extends React.Component {
  addPlaylist = playlist => {
    this.props.addPlaylist(playlist);
    this.props.history.push('/playlists');
  };
  render() {
    return (
      <div>
        <h1>เพิ่มรายการสวดมนต์</h1>
        <PlaylistForm onSubmit={this.addPlaylist} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addPlaylist: playlist => dispatch(addPlaylist(playlist))
});

export default connect(undefined, mapDispatchToProps)(AddPlaylistPage);
