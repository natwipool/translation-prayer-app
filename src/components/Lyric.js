import React from 'react';
import EachLyric from './EachLyric';

const Lyric = (props) => {
  return (
    <div>
      <h3>{props.transPrayer.precept}</h3>
      {props.transPrayer.lyrics.map((lyric, index) => 
        <EachLyric 
          key={index}
          lyric={lyric}
        />
      )}
    </div>
  )
}

export default Lyric;