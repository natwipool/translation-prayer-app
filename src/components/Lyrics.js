import React from 'react';
import Lyric from './Lyric';

const Lyrics = (props) => (
  <div>
    {
      props.transPrayers.length === 0 ? (
        <p>ไม่มีบทสวด</p>
      ) : (
        props.transPrayers.filter((transPrayer, index) => 
          index  === props.index
        ).map((result, index) => 
          <Lyric
            key={index} 
            transPrayer={result}
          />
        )
      )
    }
  </div>
)

export default Lyrics;