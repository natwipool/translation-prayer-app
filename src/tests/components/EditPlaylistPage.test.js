import React from 'react';
import { shallow } from 'enzyme';
import { EditPlaylistPage } from '../../components/EditPlaylistPage';
import playlists from '../fixtures/playlists';

test('should render EditPlaylistPage correctly', () => {
  const editPlaylist = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(
    <EditPlaylistPage
      editPlaylist={editPlaylist}
      history={history}
      playlist={playlists[0]}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should handle editPlaylist correctly', () => {
  const editPlaylist = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(
    <EditPlaylistPage
      editPlaylist={editPlaylist}
      history={history}
      playlist={playlists[0]}
    />
  );
  wrapper.find('Connect(PlaylistForm)').prop('onSubmit')(playlists[0]);
  expect(history.push).toHaveBeenCalledWith(`/playlist/${playlists[0].id}`);
  expect(editPlaylist).toHaveBeenLastCalledWith(playlists[0].id, playlists[0]);
});
