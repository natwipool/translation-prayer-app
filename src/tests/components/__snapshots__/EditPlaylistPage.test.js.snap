// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPlaylistPage correctly 1`] = `
<div>
  <Connect(PlaylistForm)
    onSubmit={[Function]}
    playlist={
      Object {
        "id": "1",
        "lists": Array [
          "คำทำวัตรเช้า",
          "ตังขณิกปัจจเวกขณปาฐะ",
          "บทพิจารณาสังขาร",
        ],
        "name": "Playlist name one",
      }
    }
  />
</div>
`;
